src/models/StyledComponent.js

```diff
@@ -29,224 +30,225 @@ type BaseState = {
-          const theme = determineTheme(
-            this.props,
-            nextTheme,
-            this.constructor.defaultProps
-          )
-          const generatedClassName = this.generateAndInjectStyles(
-            theme,
-            this.props
-          )
-
-          this.setState({ theme, generatedClassName })
-        })
-      } else {
-        // eslint-disable-next-line react/prop-types
-        const theme = this.props.theme || {}
-        const generatedClassName = this.generateAndInjectStyles(
-          theme,
-          this.props
-        )
-        this.setState({ theme, generatedClassName })
-      }
+      return className
     }
+  }
 
-    componentWillReceiveProps(nextProps: {
-      theme?: Theme,
-      [key: string]: any,
-    }) {
-      // If this is a statically-styled component, we don't need to listen to
-      // props changes to update styles
-      const { componentStyle } = this.constructor
-      if (componentStyle.isStatic) {
-        return
-      }
+  componentWillMount() {
+    const { componentStyle } = this.constructor
+    const styledContext = this.context[CHANNEL_NEXT]
 
-      this.setState(prevState => {
+    // If this is a statically-styled component, we don't need to the theme
+    // to generate or build styles.
+    if (componentStyle.isStatic) {
+      const generatedClassName = this.generateAndInjectStyles(
+        STATIC_EXECUTION_CONTEXT,
+        this.props
+      )
+      this.setState({ generatedClassName })
+      // If there is a theme in the context, subscribe to the event emitter. This
+      // is necessary due to pure components blocking context updates, this circumvents
+      // that by updating when an event is emitted
+    } else if (styledContext !== undefined) {
+      const { subscribe } = styledContext
+      this.unsubscribeId = subscribe(nextTheme => {
+        // This will be called once immediately
         const theme = determineTheme(
-          nextProps,
-          prevState.theme,
+          this.props,
+          nextTheme,
           this.constructor.defaultProps
         )
+
         const generatedClassName = this.generateAndInjectStyles(
           theme,
-          nextProps
+          this.props
         )
 
-        return { theme, generatedClassName }
+        this.setState({ theme, generatedClassName })
       })
+    } else {
+      // eslint-disable-next-line react/prop-types
+      const theme = this.props.theme || EMPTY_OBJECT
+      const generatedClassName = this.generateAndInjectStyles(theme, this.props)
+      this.setState({ theme, generatedClassName })
     }
+  }
 
-    componentWillUnmount() {
-      this.unsubscribeFromContext()
+  componentWillReceiveProps(nextProps: { theme?: Theme, [key: string]: any }) {
+    // If this is a statically-styled component, we don't need to listen to
+    // props changes to update styles
+    const { componentStyle } = this.constructor
+    if (componentStyle.isStatic) {
```

Comment:
I’m wondering whether we could eventually roll this “isStatic” check into the css helper so that we save ourselves an extra iteration